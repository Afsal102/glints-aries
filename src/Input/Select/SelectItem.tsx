import * as React from 'react';

import { SelectItemWrapper } from './SelectStyle';

const SelectItem: React.FunctionComponent<Props> = props => {
  const { children, className, onClick, ...defaultProps } = props;

  return (
    <React.Fragment>
      <SelectItemWrapper
        className={className}
        onMouseDown={onClick}
        {...defaultProps}
      >
        {children}
      </SelectItemWrapper>
    </React.Fragment>
  );
};

type Props = React.ComponentPropsWithoutRef<typeof SelectItemWrapper> & {
  children: React.ReactNode;
  onClick?(event: React.MouseEvent<HTMLLIElement, MouseEvent>): void;
  className?: string;
};

export default SelectItem;
