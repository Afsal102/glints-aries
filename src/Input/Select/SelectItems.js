/* @flow */

import React, { Fragment } from 'react';

import { Item } from '../../Style/SelectStyle';

const SelectItems = (props: Props) => {
  const {
    children,
    className,
    onClick,
    ...defaultProps
  } = props;

  return (
    <Fragment>
      <Item className={className} onMouseDown={onClick} {...defaultProps}>
        {children}
      </Item>
    </Fragment>
  );
};

type Props = {
  children: React$Node,
  onClick: Function,
  className: string,
};

export default SelectItems;
