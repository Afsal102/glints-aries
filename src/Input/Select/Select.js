import React, { Component } from 'react';

import SelectItem from './SelectItem';

import Icon from '../../General/Icon';

import {
  SelectContainer,
  SelectWrapper,
  SelectInput,
  SelectLabel,
  SelectListWrapper,
} from '../../Style/Input/SelectStyle';

class Select extends Component <Props, State> {
  state = {
    floating: false,
    isFocus: false,
    selectedValue: '',
    filterValue: [],
    cursor: 0,
    notMatch: false,
    childrenLength: 0,
    defaultValue: '',
  }

  handleFocusOut = (onBlur) => {
    const listener = (e) => {
      this.setState({
        floating: e.target.value.length > 0,
        isFocus: false,
      });

      if (onBlur !== undefined) {
        return onBlur(e);
      }
    };

    return listener;
  }

  handleFocus = (onFocus) => {
    const listener = (e) => {
      this.setState({
        isFocus: true,
      });

      if (onFocus !== undefined) {
        return onFocus(e);
      }
    };

    return listener;
  }

  handleChange = (onChange) => {
    const listener = (e) => {
      const { children } = this.props;

      this.setState({
        selectedValue: e.target.value,
        filterValue: children.filter(data => data.props.children.toLowerCase().includes(e.target.value.toLowerCase())),
        cursor: 0,
      });

      if (onChange !== undefined) {
        return onChange();
      }
    };

    return listener;
  }

  handleClick = (onOptionClick) => {
    const listener = (e) => {
      const { children, onChange } = this.props;

      this.setState({
        selectedValue: e.currentTarget.innerText,
        filterValue: children.map(data => data),
      });

      if (onChange !== undefined) {
        onChange(e.target.dataset.value);
      }

      if (onOptionClick !== undefined) {
        onOptionClick(e);
      }
    };

    return listener;
  }

  handleKeyDown = (e) => {
    const { children, onChange, disableTyping } = this.props;
    const { cursor, filterValue } = this.state;

    if (disableTyping) {
      e.preventDefault();
    }

    if (e.keyCode === 38 && cursor > 0) {
      this.setState({
        cursor: cursor - 1,
      });
    } else if (e.keyCode === 40 && cursor < filterValue.length - 1) {
      this.setState({
        cursor: cursor + 1,
      });
    } else if (e.keyCode === 13) {
      e.target.blur();
      this.setState({
        selectedValue: document.querySelector('.active').innerText,
        filterValue: children.map(data => data),
        floating: true,
        isFocus: false,
      });
      if (onChange !== undefined) {
        const itemValue = document.querySelector('.active').dataset.value;
        onChange(itemValue);
      }
    } else if (e.keyCode === 27) {
      e.target.blur();
      this.setState({ isFocus: false });
    }
  }

  handleMouseEnter = (e) => {
    this.setState({
      cursor: Number(e.target.dataset.id),
    });
  }

  componentDidMount() {
    const { value, children } = this.props;

    // Checking if children data is exist or not.
    if (children.length !== 0) {
      this.setState({
        childrenLength: children.length,
        filterValue: children.map(data => data),
      });
    } else {
      this.setState({
        notMatch: true,
      });
    }

    if (value !== undefined) {
      if (value !== '') {
        this.setState({
          floating: true,
          selectedValue: value,
          defaultValue: value,
        });
      }
    }
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.children.length !== prevState.childrenLength) {
      return {
        filterValue: nextProps.children.map(data => data),
        childrenLength: nextProps.children.length,
        selectedValue: nextProps.value,
        defaultValue: nextProps.value,
      };
    }

    if (prevState.filterValue.length === 0) {
      return { notMatch: true };
    }

    if (nextProps.value !== undefined && nextProps.value !== prevState.defaultValue) {
      return {
        selectedValue: nextProps.value,
        defaultValue: nextProps.value,
      };
    }

    return { notMatch: false };
  }

  render() {
    const {
      label,
      value,
      status,
      disabled,
      className,
      onFocus,
      onBlur,
      onChange,
      noOptionResult,
      children,
      small,
      disableTyping,
      removeFloatingLabel,
      removeDropIcon,
      ...defaultProps
    } = this.props;

    const {
      floating,
      isFocus,
      selectedValue,
      filterValue,
      cursor,
      notMatch,
    } = this.state;

    return (
      <SelectContainer id="aries-select" className={className}>
        <SelectWrapper id="select-inputwrapper" isFocus={isFocus}>
          <SelectInput
            type="text"
            placeholder={removeFloatingLabel && label}
            role="combobox"
            aria-expanded={isFocus}
            aria-autocomplete="list"
            status={status}
            disabled={disabled}
            onFocus={this.handleFocus(onFocus)}
            onBlur={this.handleFocusOut(onBlur)}
            onChange={this.handleChange(onChange)}
            onKeyDown={this.handleKeyDown}
            floating={floating}
            value={selectedValue}
            small={small}
            disableTyping={disableTyping}
            {...defaultProps}
          />
          <If condition={!removeFloatingLabel}>
            <SelectLabel
              floating={floating}
              status={status}
              small={small}
            >
              {label}
            </SelectLabel>
          </If>
          <If condition={!removeDropIcon}>
            <div
              className="select-icon"
              aria-label="show options"
            >
              <Icon name="arrow-down" color="#777777" />
            </div>
          </If>
        </SelectWrapper>
        <SelectListWrapper
          id="select-listbox"
          role="listbox"
          aria-hidden={!isFocus && true}
          open={isFocus}
          small={small}
        >
          {filterValue.map((data, index) => (
            <SelectItem
              className={cursor === index ? 'active' : null}
              key={data.props.value}
              role="option"
              data-id={index}
              data-value={data.props.value}
              onClick={this.handleClick(data.props.onOptionClick)}
              onMouseEnter={this.handleMouseEnter}
              tabIndex="0"
            >
              {data.props.children}
            </SelectItem>
          ))}
          {notMatch && (
            <SelectItem
              disabled
              role="option"
              aria-hidden={false}
              aria-disabled="true"
            >
              {noOptionResult}
            </SelectItem>
          )}
        </SelectListWrapper>
      </SelectContainer>
    );
  }
}

type Props = {
  label: string,
  status: string,
  disabled: boolean,
  className: string,
  noOptionResult: string,
  children: React$Node,
  small: boolean,
  disableTyping: boolean,
  removeFloatingLabel: boolean,
  removeDropIcon: boolean,
  defaultValue: string,
}

type State = {
  floating: boolean,
  isFocus: boolean,
  selectedValue: string,
  filterValue: array,
  cursor: number,
  notMatch: boolean,
  childrenLength: number,
};

export default Select;
