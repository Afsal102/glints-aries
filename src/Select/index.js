// @flow

import Select from './Select';
import SelectItems from './SelectItems';

Select.Item = SelectItems;

export default Select;
