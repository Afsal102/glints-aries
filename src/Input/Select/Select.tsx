import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { get, toLower } from 'lodash';

import classNames from 'classnames';

import SelectList, { SelectItemProps } from './SelectList';

import { ArrowDownIcon } from '../../General/Icon/components';
import {
  SelectContainer,
  SelectWrapper,
  SelectInput,
  SelectLabel,
  SelectErrorDefault,
} from '../../Style/Input/SelectStyle';

const Select: ISelect = (props: Props) => {
  const {
    label,
    status,
    disabled,
    className,
    onFocus,
    onBlur,
    onChange,
    onInputChange,
    noOptionResult = 'No results found',
    small,
    disableTyping,
    removeFloatingLabel,
    removeDropIcon,
    error,
    renderError,
    value,
    defaultValue,
    defaultOpen = false,
    children,
    isLoading,
    ...defaultProps
  } = props;

  // warn deprecated prop "status"
  React.useEffect(
    function warnDeprecatedProp() {
      if (status) {
        if (typeof console !== 'undefined') {
          console.warn(`Warning: Select's status prop is deprecated and will be
        removed in a future release.\n\nPlease use the error prop instead to
        show errors and indicate an error state.`);
        }
      }
    },
    [status]
  );

  const [floating, setFloating] = React.useState<boolean>(false);
  const [isFocus, setIsFocus] = React.useState<boolean>(defaultOpen);
  const [isInputChange, setIsInputChange] = React.useState<boolean>(false);
  const [inputValue, setInputValue] = React.useState<string>(
    value || defaultValue || ''
  );
  const [activeOptionIndex, setActiveOptionIndex] = React.useState<number>(0);
  const [
    shouldScrollToActiveOption,
    setShouldScrollToActiveOption,
  ] = React.useState<boolean>(false);

  const selectContainerRef: React.RefObject<HTMLDivElement> = React.useRef();
  const selectInputRef: React.RefObject<HTMLInputElement> = React.useRef();

  // set options based on children and inputValue
  const availableOptions = React.useMemo(() => {
    const childrenOptions = React.Children.toArray(children) as Array<
      React.ReactElement<SelectItemProps>
    >;
    if (!inputValue) {
      return childrenOptions;
    }

    if (inputValue && disableTyping) {
      return childrenOptions;
    }

    if (!isInputChange) {
      const isInputValueOneOfOptions = childrenOptions.some(
        data => data.props.children === inputValue
      );
      if (isInputValueOneOfOptions) {
        return childrenOptions;
      }
    }

    const matchedChildrenOptions = childrenOptions.filter(data =>
      toLower(data.props.children).includes(toLower(inputValue))
    );
    return matchedChildrenOptions;
  }, [children, inputValue, disableTyping, isInputChange]);

  const [options, setOptions] = React.useState<React.ReactNode[]>(
    availableOptions
  );

  React.useEffect(
    function updateOptions() {
      setOptions(availableOptions);
    },
    [availableOptions]
  );

  React.useEffect(function registerClickOutsideListener() {
    const onClickOutside = (event: MouseEvent) => {
      const element = event.target as HTMLElement;
      if (
        selectContainerRef.current &&
        !ReactDOM.findDOMNode(selectContainerRef.current).contains(element)
      ) {
        setIsFocus(false);
      }
    };
    document.addEventListener('click', onClickOutside, false);
    return () => document.removeEventListener('click', onClickOutside, false);
  }, []);

  React.useEffect(
    function handleValueChange() {
      if (value) {
        setFloating(true);
        setInputValue(value as string);
        return;
      }
      if (value === '') {
        setInputValue(value);
        setFloating(false);
      }
    },
    [value]
  );

  const handleFocusOut = React.useCallback(
    (e: React.FocusEvent<HTMLInputElement>) => {
      setFloating(e.target.value.length > 0);
      setIsInputChange(false);
      if (typeof onBlur === 'function') {
        return onBlur(e);
      }
    },
    [onBlur]
  );

  const handleFocus = React.useCallback(
    (e: React.FocusEvent<HTMLInputElement>) => {
      setIsFocus(true);

      if (typeof onFocus === 'function') {
        return onFocus(e);
      }
    },
    [onFocus]
  );

  const handleClick = () => {
    setIsFocus(!isFocus);
  };

  const handleDropIconClick = () => {
    if (!disableTyping && !isFocus) {
      selectInputRef.current.focus();
    }
    setIsFocus(!isFocus);
  };

  // Should be called when the user types into the input
  const handleInputChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(e.target.value);
      setActiveOptionIndex(0);
      setIsInputChange(true);

      if (
        typeof onChange === 'function' &&
        typeof onInputChange !== 'function'
      ) {
        onChange(e);
        if (typeof console !== 'undefined') {
          console.warn(`
            Warning: onChange will not be fired when input value changes in a future release,
            please use onInputChange instead. Now, if onChange is passed but onInputChange is not,
            the onChange callback will still receive the input change event for backward compatibility.
          `);
        }
      }

      if (typeof onInputChange === 'function') {
        onInputChange(e);
      }
    },
    [onChange, onInputChange]
  );

  const getActiveElement = React.useCallback(() => {
    return selectContainerRef.current.querySelector('.active') as HTMLLIElement;
  }, []);

  // Should be called when the user selects an option
  const handleClickOnOption = React.useCallback(
    (e?: React.ChangeEvent<HTMLInputElement>) => {
      const activeElement = e ? e.target : getActiveElement();
      const inputValue = activeElement.textContent;
      setInputValue(inputValue);
      setIsFocus(false);
      setFloating(true);
      setIsInputChange(false);
      if (typeof onChange === 'function') {
        onChange(activeElement.dataset.value);
      }

      const activeElementIndex = Number(get(activeElement, 'dataset.id'));
      const activeChild = options[activeElementIndex];
      const onOptionClick = get(activeChild, 'props.onOptionClick');
      if (typeof onOptionClick === 'function') {
        onOptionClick(e);
      }
    },
    [onChange, options, getActiveElement]
  );

  const scrollToActiveElement = React.useCallback(() => {
    const selectListElement = selectContainerRef.current.querySelector(
      '.select-listbox'
    ) as HTMLUListElement;
    selectListElement.scrollTop = getActiveElement().offsetTop;
  }, [getActiveElement]);

  // Should be called when any key is pressed inside the component
  const handleKeyDown = React.useCallback(
    (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (disableTyping) {
        e.preventDefault();
      }

      const inputElement = e.target as HTMLInputElement;

      // up arrow key
      if (e.keyCode === 38 && activeOptionIndex > 0) {
        setActiveOptionIndex(activeOptionIndex - 1);
        setShouldScrollToActiveOption(true);
      }
      // down arrow key
      else if (e.keyCode === 40 && activeOptionIndex < options.length - 1) {
        setActiveOptionIndex(activeOptionIndex + 1);
        setShouldScrollToActiveOption(true);
      }
      // enter key
      else if (e.keyCode === 13) {
        inputElement.blur();
        handleClickOnOption();
      }
      // escape key
      else if (e.keyCode === 27) {
        inputElement.blur();
        setIsFocus(false);
      }
    },
    [disableTyping, options, activeOptionIndex, handleClickOnOption]
  );

  React.useEffect(
    function handleScrollToActiveElement() {
      if (shouldScrollToActiveOption) {
        scrollToActiveElement();
        setShouldScrollToActiveOption(false);
      }
    },
    [
      shouldScrollToActiveOption,
      setShouldScrollToActiveOption,
      scrollToActiveElement,
    ]
  );

  const handleMouseEnterOption = React.useCallback(
    (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
      const listItemElement = e.target as HTMLLIElement;
      setActiveOptionIndex(Number(listItemElement.dataset.id));
    },
    []
  );

  const deprecatedStatus = status || (error && 'error');

  const shouldShowError = error && typeof error !== 'boolean';

  let completeError = null;
  if (shouldShowError) {
    if (renderError) {
      completeError = renderError(error);
    } else {
      completeError = (
        <SelectErrorDefault className="select-error-default">
          {error}
        </SelectErrorDefault>
      );
    }
  }

  return (
    <SelectContainer
      className={classNames('aries-select', className)}
      ref={selectContainerRef}
    >
      <SelectWrapper
        className="select-inputwrapper"
        isFocus={isFocus}
        small={small}
      >
        <SelectInput
          ref={selectInputRef}
          type="text"
          placeholder={removeFloatingLabel && label}
          role="combobox"
          aria-expanded={isFocus}
          aria-autocomplete="list"
          status={deprecatedStatus}
          disabled={disabled}
          onFocus={disableTyping ? null : handleFocus}
          onBlur={disableTyping ? null : handleFocusOut}
          onChange={handleInputChange}
          onKeyDown={handleKeyDown}
          onClick={disableTyping ? handleClick : null}
          floating={floating}
          value={inputValue}
          small={small}
          disableTyping={disableTyping}
          readOnly={disableTyping}
          autoFocus={defaultOpen}
          {...defaultProps}
        />
        {!removeFloatingLabel && (
          <SelectLabel
            aria-label={label}
            data-testid="select-label"
            floating={floating}
            status={deprecatedStatus}
            small={small}
          >
            {label}
          </SelectLabel>
        )}
        {!removeDropIcon && (
          <div
            className="select-icon"
            aria-label="show options"
            onClick={handleDropIconClick}
          >
            <ArrowDownIcon color="#777777" />
          </div>
        )}
      </SelectWrapper>
      <SelectList
        aria-label="select-list"
        activeOptionIndex={activeOptionIndex}
        options={options}
        isFocus={isFocus}
        isLoading={isLoading}
        noOptionResult={noOptionResult}
        small={small}
        handleClickOnOption={handleClickOnOption}
        handleMouseEnterOption={handleMouseEnterOption}
      />
      {completeError}
    </SelectContainer>
  );
};

interface Props extends React.ComponentPropsWithoutRef<typeof SelectInput> {
  children: React.ReactNode;
  isLoading?: boolean;
  label?: string;
  noOptionResult?: string;
  removeDropIcon?: boolean;
  removeFloatingLabel?: boolean;
  error?: React.ReactNode | string | boolean;
  renderError?: (error: React.ReactNode | string | boolean) => React.ReactNode;
  defaultValue?: string;
  defaultOpen?: boolean;

  onFocus?(e: React.FocusEvent<HTMLInputElement>): void;
  onBlur?(e: React.FocusEvent<HTMLInputElement>): void;
  onChange?(value: any): void;
  onInputChange?(e: React.ChangeEvent<HTMLInputElement>): void;
}

type ISelect = React.FunctionComponent<Props> & {
  Option: React.FunctionComponent<SelectItemProps>;
};

export const Option: React.FunctionComponent<SelectItemProps> = () => null;
Option.displayName = 'Select.Option';

Select.Option = Option;

export default Select;
