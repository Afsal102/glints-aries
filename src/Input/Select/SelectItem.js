/* @flow */

import React, { Fragment } from 'react';

import { SelectItemWrapper } from '../../Style/Input/SelectStyle';

const SelectItem = (props: Props) => {
  const {
    children,
    className,
    onClick,
    ...defaultProps
  } = props;

  return (
    <Fragment>
      <SelectItemWrapper className={className} onMouseDown={onClick} {...defaultProps}>
        {children}
      </SelectItemWrapper>
    </Fragment>
  );
};

type Props = {
  children: React$Node,
  onClick: Function,
  className: string,
};

export default SelectItem;