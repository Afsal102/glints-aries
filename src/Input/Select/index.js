// @flow

import Select from './Select';
import SelectItem from './SelectItem';

Select.Option = SelectItem;

export default Select;
