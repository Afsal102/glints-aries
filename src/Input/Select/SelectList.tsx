import * as React from 'react';

import SelectItem from './SelectItem';

import Loading from '../../General/Loading';
import { SelectListWrapper } from '../../Style/Input/SelectStyle';

const SelectList: React.FunctionComponent<Props> = ({
  cursor,
  filterValue,
  isFocus,
  isLoading,
  noOptionResult,
  small,
  handleClick,
  handleMouseEnter,
}: Props) => (
  <SelectListWrapper
    className="select-listbox"
    role="listbox"
    data-testid="listbox"
    aria-hidden={!isFocus && true}
    open={isFocus}
    small={small}
  >
    {filterValue.length !== 0 && !isLoading ? (
      filterValue.map((data: React.ReactElement<SelectItemProps>, index) => (
        <SelectItem
          className={cursor === index ? 'active' : null}
          key={data.props.value}
          role="option"
          data-testid="option"
          data-id={index}
          data-value={data.props.value}
          onClick={handleClick}
          onMouseEnter={handleMouseEnter}
          tabIndex={0}
        >
          {data.props.children}
        </SelectItem>
      ))
    ) : isLoading ? (
      <SelectItem className="select-loading" role="option">
        <Loading />
      </SelectItem>
    ) : (
      <SelectItem
        disabled
        role="option"
        data-testid="option"
        aria-hidden={false}
        aria-disabled="true"
      >
        {noOptionResult}
      </SelectItem>
    )}
  </SelectListWrapper>
);

interface Props {
  cursor: number;
  filterValue: React.ReactNode[];
  isFocus: boolean;
  isLoading: boolean;
  noOptionResult?: string;
  small?: boolean;
  handleClick?: () => void;
  handleMouseEnter(event: React.MouseEvent<HTMLLIElement, MouseEvent>): void;
}

export interface SelectItemProps {
  children: string;
  value?: string;
  onOptionClick?(e: React.MouseEvent<HTMLLIElement, MouseEvent>): void;
}

export default SelectList;
