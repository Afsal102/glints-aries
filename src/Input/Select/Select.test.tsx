import * as React from 'react';

import * as lodash from 'lodash';
import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Select from './Select';
import { PrimaryColor } from '../../Style/Colors';

const props = {
  label: 'Jobs',
  onChange: jest.fn(value => value),
  onOptionClick: jest.fn(),
  options: [
    { id: '1', value: 'accountant', name: 'Accountant' },
    { id: '2', value: 'business development', name: 'Business Development' },
    { id: '3', value: 'software engineer', name: 'Software Engineer' },
  ],
};

const SelectChildren = props.options.map(option => (
  <Select.Option
    key={option.id}
    value={option.value}
    onOptionClick={props.onOptionClick}
  >
    {option.name}
  </Select.Option>
));

const SelectComponent = (
  <Select label={props.label} onChange={props.onChange}>
    {SelectChildren}
  </Select>
);

function setupOpenSelectMenu() {
  const { getByRole, getAllByRole } = render(SelectComponent);
  const selectInput = getByRole('combobox') as HTMLInputElement;
  const selectList = getByRole('listbox');
  const displayedOptions = getAllByRole('option');

  fireEvent.focus(selectInput);

  return {
    selectInput,
    selectList,
    displayedOptions,
  };
}

function setupSelectOptionFromMenu() {
  const { selectInput, selectList, displayedOptions } = setupOpenSelectMenu();
  const randomOption =
    displayedOptions[Math.floor(Math.random() * displayedOptions.length)];

  fireEvent.mouseDown(randomOption);
  fireEvent.change(selectInput, {
    target: { value: randomOption.textContent },
  });

  return {
    selectInput,
    selectList,
    randomOption,
  };
}

function setupSelectOptionFromMenuWithKeyHandling() {
  const { selectInput, selectList, displayedOptions } = setupOpenSelectMenu();
  const randomIndex = Math.floor(Math.random() * displayedOptions.length);
  const randomOption = displayedOptions[randomIndex];
  const arrowDownEvent = { key: 'ArrowDown', keyCode: 40 };
  const enterEvent = { key: 'Enter', keyCode: 13 };

  lodash.times(randomIndex, () =>
    fireEvent.keyDown(selectInput, arrowDownEvent)
  );
  fireEvent.keyDown(selectInput, enterEvent);
  fireEvent.change(selectInput, {
    target: { value: randomOption.textContent },
  });

  return {
    selectInput,
    selectList,
    randomOption,
  };
}

afterEach(() => {
  jest.clearAllMocks();
});

it('<Select> should render with 3 options', () => {
  const SelectSnapshot = renderer.create(SelectComponent).toJSON();
  expect(SelectSnapshot).toMatchSnapshot();
});

describe('when it is rendered', () => {
  it('should display the correct input label', () => {
    const { getByLabelText } = render(SelectComponent);
    const selectLabel = getByLabelText(props.label);
    expect(selectLabel).toHaveTextContent(props.label);
  });

  it('should display an empty input', async () => {
    const { getByRole } = render(SelectComponent);
    const selectInput = getByRole('combobox');
    expect(selectInput).toHaveTextContent('');
  });

  it('should not render the default error if error is falsy', () => {
    const { getByRole } = render(<Select>{SelectChildren}</Select>);
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    expect(selectContainer.lastChild).not.toHaveClass('select-error-default');
  });

  it('should not render the default error if error is true', () => {
    const { getByRole } = render(
      <Select error={true}>{SelectChildren}</Select>
    );
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    expect(selectContainer.lastChild).not.toHaveClass('select-error-default');
  });

  it('should render the error string in the default error container', () => {
    const { getByRole } = render(
      <Select error={'test'}>{SelectChildren}</Select>
    );
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    expect(selectContainer.lastChild).toHaveClass('select-error-default');
    expect(selectContainer.lastChild).toHaveTextContent('test');
  });

  it('should render the error component in the default error container', () => {
    const customError = <span data-testid="custom-error">nasty error</span>;
    const { container, getByTestId } = render(
      <Select error={customError}>{SelectChildren}</Select>
    );
    const customErrorElement = getByTestId('custom-error');
    expect(container).toContainElement(customErrorElement);
    expect(customErrorElement).toHaveTextContent('nasty error');
  });

  it('should render the error string in the custom error', () => {
    const customError = (e: string) => (
      <div data-testid="custom-error">{e}</div>
    );
    const { container, getByTestId } = render(
      <Select renderError={customError} error="nasty error">
        {SelectChildren}
      </Select>
    );
    const errorElement = getByTestId('custom-error');
    expect(container).toContainElement(errorElement);
    expect(container).toHaveTextContent('nasty error');
  });
});

describe('when it is clicked', () => {
  it('should open the option menu', () => {
    const { selectList } = setupOpenSelectMenu();
    expect(selectList.hasAttribute('open')).toEqual(true);
  });

  it('should display all options with the correct value', () => {
    const { displayedOptions } = setupOpenSelectMenu();
    displayedOptions.forEach((option, index) => {
      expect(option).toHaveTextContent(props.options[index].name);
      expect(option.getAttribute('data-value')).toEqual(
        props.options[index].value
      );
    });
  });
});

describe('when an option is clicked', () => {
  it('should call the onOptionClick function', () => {
    setupSelectOptionFromMenu();
    expect(props.onOptionClick).toHaveBeenCalledTimes(1);
  });

  it('should show the value of the option on the select input', () => {
    const { selectInput, randomOption } = setupSelectOptionFromMenu();
    expect(selectInput.value).toEqual(randomOption.textContent);
  });

  it('should close the option menu', () => {
    const { selectList } = setupSelectOptionFromMenu();
    expect(selectList.hasAttribute('open')).toBe(false);
  });
});

describe('when the enter key is pressed on an option', () => {
  it('should call the onOptionClick function', () => {
    setupSelectOptionFromMenuWithKeyHandling();
    expect(props.onOptionClick).toHaveBeenCalledTimes(1);
  });

  it('should show the value of the option on the select input', () => {
    const {
      selectInput,
      randomOption,
    } = setupSelectOptionFromMenuWithKeyHandling();
    expect(selectInput.value).toEqual(randomOption.textContent);
  });

  it('should close the option menu', () => {
    const { selectList } = setupSelectOptionFromMenuWithKeyHandling();
    expect(selectList.hasAttribute('open')).toBe(false);
  });
});

describe('when no results are found', () => {
  it('should display "No results found"', () => {
    const { selectInput, selectList } = setupOpenSelectMenu();
    fireEvent.change(selectInput, {
      target: { value: 'z' },
    });
    expect(selectList).toHaveTextContent(/^No results found$/);
  });
});

describe("when status = 'error'", () => {
  it('should show a red border', () => {
    const { getByRole } = render(
      <Select status="error">{SelectChildren}</Select>
    );
    const selectInput = getByRole('combobox');
    expect(selectInput).toHaveStyle(`border-color: ${PrimaryColor.glintsred}`);
  });
});

describe('when error is true', () => {
  it('should show a red border', () => {
    const { getByRole } = render(
      <Select error={true}>{SelectChildren}</Select>
    );
    const selectInput = getByRole('combobox');
    expect(selectInput).toHaveStyle(`border-color: ${PrimaryColor.glintsred}`);
  });
});

describe('when error is a string', () => {
  it('should show a red border', () => {
    const { getByRole } = render(
      <Select error="error">{SelectChildren}</Select>
    );
    const selectInput = getByRole('combobox');
    expect(selectInput).toHaveStyle(`border-color: ${PrimaryColor.glintsred}`);
  });
});

describe('when error is a component', () => {
  it('should show a red border', () => {
    const { getByRole } = render(
      <Select error={<span>error</span>}>{SelectChildren}</Select>
    );
    const selectInput = getByRole('combobox');
    expect(selectInput).toHaveStyle(`border-color: ${PrimaryColor.glintsred}`);
  });
});

describe('when renderError is given', () => {
  it('should not render the custom error if error is falsy', () => {
    const customError = (e: string) => <div className="customError">{e}</div>;
    const { getByRole } = render(
      <Select renderError={customError}>{SelectChildren}</Select>
    );
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    expect(selectContainer.lastChild).not.toHaveClass('customError');
  });

  it('should not render the custom error if error is true', () => {
    const customError = (e: string) => <div className="customError">{e}</div>;
    const { getByRole } = render(
      <Select renderError={customError} error={true}>
        {SelectChildren}
      </Select>
    );
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    expect(selectContainer.lastChild).not.toHaveClass('customError');
  });

  it('should render the error string in the custom error container', () => {
    const errorContainer = (e: React.ReactNode) => (
      <div data-testid="error-component">{e}</div>
    );
    const { container, getByTestId } = render(
      <Select renderError={errorContainer} error="nasty error">
        {SelectChildren}
      </Select>
    );
    const errorContainerElement = getByTestId('error-component');
    expect(container).toContainElement(errorContainerElement);
    expect(errorContainerElement).toHaveTextContent('nasty error');
  });

  it('should render the custom error in the custom error container', () => {
    const customError = () => (
      <span data-testid="custom-error">nasty error</span>
    );
    const errorContainer = (e: () => React.ReactNode) => (
      <div data-testid="error-component">{e()}</div>
    );
    const { container, getByTestId } = render(
      <Select renderError={errorContainer} error={customError}>
        {SelectChildren}
      </Select>
    );
    const customErrorElement = getByTestId('custom-error');
    const errorContainerElement = getByTestId('error-component');
    expect(container).toContainElement(errorContainerElement);
    expect(errorContainerElement).toContainElement(customErrorElement);
    expect(customErrorElement).toHaveTextContent('nasty error');
  });
});

describe('when isLoading = true', () => {
  it('should show a loading spinner', () => {
    const { getByRole } = render(<Select isLoading>{SelectChildren}</Select>);
    const selectContainer = getByRole('combobox').parentElement.parentElement;
    const loadingSpinner = getByRole('alert');
    expect(selectContainer).toContainElement(loadingSpinner);
  });
});

it('up and down arrow keys can be used to navigate the options menu', () => {
  const { selectInput, displayedOptions } = setupOpenSelectMenu();
  const firstOption = displayedOptions[0];
  const secondOption = displayedOptions[1];
  const arrowUpEvent = { key: 'ArrowUp', keyCode: 38 };
  const arrowDownEvent = { key: 'ArrowDown', keyCode: 40 };

  expect(firstOption).toHaveClass('active');

  fireEvent.keyDown(selectInput, arrowDownEvent);
  expect(secondOption).toHaveClass('active');

  fireEvent.keyDown(selectInput, arrowUpEvent);
  expect(firstOption).toHaveClass('active');
});
