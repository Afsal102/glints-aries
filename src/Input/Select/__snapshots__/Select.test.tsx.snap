// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select> should render with 3 options 1`] = `
<div
  className="aries-select SelectStyle__SelectContainer-jmk38c-0 cxQAyJ"
>
  <div
    className="select-inputwrapper SelectStyle__SelectWrapper-jmk38c-1 cnQpwy"
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      className="SelectStyle__SelectInput-jmk38c-3 bGZcBu"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="combobox"
      type="text"
      value=""
    />
    <label
      aria-label="Jobs"
      className="SelectStyle__SelectLabel-jmk38c-2 fuofeP"
    >
      Jobs
    </label>
    <div
      aria-label="show options"
      className="select-icon"
    >
      <svg
        data-testid="icon-svg"
        fill="#777777"
        height="1em"
        viewBox="0 0 100 100"
        width="1em"
      >
        <polygon
          points="11.7984725 19 49.9935275 57.3950291 88.1885825 19 99.987055 30.7984725 49.9935275 80.792 0 30.7984725"
        />
      </svg>
    </div>
  </div>
  <ul
    aria-hidden={true}
    className="select-listbox SelectStyle__SelectListWrapper-jmk38c-4 gVUTIk"
    open={false}
    role="listbox"
  >
    <li
      className="active SelectStyle__SelectItemWrapper-jmk38c-5 bTfIVb"
      data-id={0}
      data-value="accountant"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      role="option"
      tabIndex={0}
    >
      Accountant
    </li>
    <li
      className="SelectStyle__SelectItemWrapper-jmk38c-5 bTfIVb"
      data-id={1}
      data-value="business development"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      role="option"
      tabIndex={0}
    >
      Business Development
    </li>
    <li
      className="SelectStyle__SelectItemWrapper-jmk38c-5 bTfIVb"
      data-id={2}
      data-value="software engineer"
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      role="option"
      tabIndex={0}
    >
      Software Engineer
    </li>
  </ul>
</div>
`;
